subsys/bluetooth/lib/ead.c

```diff
@@ -0,0 +1,220 @@
+
+	return data_len + 1;
+}
+
+static int ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		       const uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE], const uint8_t *payload,
+		       size_t payload_size, uint8_t *encrypted_payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	size_t ead_size = BT_EAD_RANDOMIZER_SIZE + payload_size + BT_EAD_MIC_SIZE;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
+	if (err != 0) {
+		return -ECANCELED;
+	}
+
+	memcpy(encrypted_payload, nonce, BT_EAD_RANDOMIZER_SIZE);
+
+	err = bt_ccm_encrypt(session_key, nonce, payload, payload_size, bt_ead_aad, BT_EAD_AAD_SIZE,
+			     &encrypted_payload[BT_EAD_RANDOMIZER_SIZE], BT_EAD_MIC_SIZE);
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
+
+	LOG_HEXDUMP_DBG(encrypted_payload, ead_size, "Encrypted Data: ");
+
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		   const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
+
+	return ead_encrypt(session_key, iv, NULL, payload, payload_size, encrypted_payload);
+}
+
+#if defined(CONFIG_BT_TESTING)
+
+int bt_test_ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE],
+			const uint8_t iv[BT_EAD_IV_SIZE],
+			const uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE], const uint8_t *payload,
+			size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
+
+	return ead_encrypt(session_key, iv, randomizer, payload, payload_size, encrypted_payload);
+}
+
+#endif /* CONFIG_BT_TESTING */
+
+static int ead_decrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		       const uint8_t *encrypted_payload, size_t encrypted_payload_size,
+		       uint8_t *payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	const uint8_t *encrypted_ad_data = &encrypted_payload[BT_EAD_RANDOMIZER_SIZE];
+	size_t encrypted_ad_data_size = encrypted_payload_size - BT_EAD_RANDOMIZER_SIZE;
+	size_t payload_size = encrypted_ad_data_size - BT_EAD_MIC_SIZE;
+
+	const uint8_t *randomizer = encrypted_payload;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
+	if (err != 0) {
+		return -EIO;
+	}
+
+	LOG_HEXDUMP_DBG(encrypted_ad_data, encrypted_ad_data_size, "Encrypted Data: ");
```

Comment:
You're right for decryption, it can be useful to log the decrypted data before returning. For encryption it's not useful, as discussed with @Thalley previously the `bt_ccm_encrypt` return only before doing any encryption, at the args check.